Support DI sharding for FPE_EBC

hannaxu opened the pull request:
Summary:
Support models that have FeatureProcessedEmbeddingBagCollection.

These changes make sure we add QuantFeatureProcessedEmbeddingBagCollectionSharder as a recognized sharder, handle multiple envs needed for specifying DI sharding, and propagate TBE properly when processing the sharding plan.

Differential Revision: D74671655

jingsh commented on torchrec/distributed/quant_embeddingbag.py:
```diff
@@ -462,11 +462,30 @@ def __init__(
+                assert (
+                    shard_device_type == expected_device_type
+                ), f"Expected {expected_device_type} but got {shard_device_type} for FeatureProcessedEmbeddingBagCollection sharding device type"
+
```
since we assert here and the `expected_device_type` is always `cuda`, why don't we keep the signature of the method's env as a single `ShardingEnv`, and just pass the `cuda`'s sharding evn to this?

hannaxu replied:
While right now FPE is used on cuda, we can support it on cuda/cpu in the future. Performing the check here allows us to de-duplicate the logic so we can ensure there's one sharding type for FPE for all sharding paths.

Commit:
(1/n) Allow DI sharding for models with FPE_EBC (#2968)

Summary:
Pull Request resolved: https://github.com/pytorch/torchrec/pull/2968

Support models that have FeatureProcessedEmbeddingBagCollection to be DI sharded. However, conservatively enforce that FPE itself can only be sharded on HBM and not across CPU as well.

These changes make sure we add QuantFeatureProcessedEmbeddingBagCollectionSharder as a recognized sharder, handle multiple envs needed for specifying DI sharding, and propagate TBE properly when processing the sharding plan.

This doesn't support true hybrid sharding for FPE.

Reviewed By: faran928

Differential Revision: D74671655
